File: weird_captcha_gym/shared_runtime/app/mechanics/consent_gauntlet.js
(() => {
  "use strict";

  let model = null;
  let cleanup = null;
  const esc = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const interaction = () => model.state.control_condition?.interaction || "full";
  const parameter = (key) => model.state.parameters?.[key];

  function stopMotion() {
    if (model?.motionFrame) cancelAnimationFrame(model.motionFrame);
    if (model) model.motionFrame = 0;
  }

  function record(event) {
    model.events.push({
      sequence: model.events.length + 1,
      task_time_ms: Number(Math.max(0, performance.now() - model.runStartedAt).toFixed(3)),
      ...event,
    });
    if (model.freshFailure) {
      model.freshFailure = false;
      const verdict = document.querySelector(".consent-verdict");
      if (verdict) verdict.className = "consent-verdict";
      model.helpers.setReadout("PRIVACY PACKET ACTIVE", "idle");
    }
  }

  function optionById(id, stage = model.stage) {
    const key = stage === "entry" ? "entry_options" : "final_options";
    return model.surface[key].find((item) => item.id === id) || null;
  }

  function drawerById(id) {
    return model.surface.drawers.find((item) => item.id === id) || null;
  }

  function purposeById(id) {
    return model.surface.purposes.find((item) => item.id === id) || null;
  }

  function finalState() {
    return {
      stage: model.stage,
      current_drawer: model.currentDrawer,
      purpose_states: Object.fromEntries(model.surface.purposes.map((item) => [item.id, item.state])),
    };
  }

  function showVerdict(kind) {
    const verdict = document.querySelector(".consent-verdict");
    if (!verdict) return;
    verdict.className = `consent-verdict is-${kind}`;
    verdict.innerHTML = `<small>PRIVACY PACKET</small><b>${kind.toUpperCase()}</b>`;
  }

  async function submit(completed) {
    if (!model || model.submitting || model.terminal) return;
    const current = model;
    current.submitting = true;
    current.helpers.setReadout("CHECKING CONSENT LEDGER…", "pending");
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify({
          mechanic_id: current.state.mechanic_id,
          task_id: current.state.task_id,
          challenge_id: current.state.challenge_id,
          interaction_mode: interaction(),
          events: current.events,
          final_state: finalState(),
          elapsed_task_ms: Number(Math.max(0, performance.now() - current.runStartedAt).toFixed(3)),
          completed: completed === true,
        }),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        current.terminal = true;
        stopMotion();
        current.helpers.setReadout("PASS", "passed");
        showVerdict("pass");
      } else if (outcome.passed === false && outcome.state) {
        const helpers = current.helpers;
        await render(outcome.state, helpers, {freshFailure: true});
        model.helpers.setReadout("FAIL · NEW PACKET LOADED", "error");
        showVerdict("fail");
      } else {
        current.submitting = false;
        current.helpers.setReadout("LEDGER REJECTED", "error");
      }
    } catch (_error) {
      if (model === current) {
        current.submitting = false;
        current.helpers.setReadout("PRIVACY LINK OFFLINE", "error");
      }
    }
  }

  function optionClick(event, option) {
    if (model.submitting || model.terminal) return;
    const source = interaction() === "full" ? "orbit_card" : "option_proxy";
    const proof = {};
    if (interaction() === "full") {
      const rect = event.currentTarget.getBoundingClientRect();
      const orbitRect = event.currentTarget.closest(".consent-orbit").getBoundingClientRect();
      const centerX = rect.left + rect.width / 2;
      const centerY = rect.top + rect.height / 2;
      const dx = (event.clientX - (rect.left + rect.width / 2)) / Math.max(1, rect.width / 2);
      const dy = (event.clientY - (rect.top + rect.height / 2)) / Math.max(1, rect.height / 2);
      proof.pointer_offset_norm = Number(Math.hypot(dx, dy).toFixed(5));
      proof.phase_deg = Number(currentPhase().toFixed(5));
      proof.pointer_x_norm = Number(((event.clientX - orbitRect.left) / orbitRect.width).toFixed(6));
      proof.pointer_y_norm = Number(((event.clientY - orbitRect.top) / orbitRect.height).toFixed(6));
      proof.card_center_x_norm = Number(((centerX - orbitRect.left) / orbitRect.width).toFixed(6));
      proof.card_center_y_norm = Number(((centerY - orbitRect.top) / orbitRect.height).toFixed(6));
      proof.card_width_norm = Number((rect.width / orbitRect.width).toFixed(6));
      proof.card_height_norm = Number((rect.height / orbitRect.height).toFixed(6));
    }
    record({type: "gateway", id: option.id, input_source: source, ...proof});
    if (model.stage === "entry") {
      if (option.action === "manage") {
        model.stage = "preferences";
        renderStage();
      } else {
        submit(false);
      }
    } else if (option.action === "commit") {
      submit(true);
    } else {
      submit(false);
    }
  }

  function currentPhase() {
    const elapsed = (performance.now() - model.motionStart) / 1000;
    return model.motionPhase + (parameter("moving_gateways") ? elapsed * parameter("orbit_speed_deg_per_second") : 0);
  }

  function positionOrbit() {
    const elapsed = Math.max(0, performance.now() - model.runStartedAt);
    const timer = document.querySelector("[data-packet-time]");
    if (timer) {
      const minutes = Math.floor(elapsed / 60000);
      const seconds = ((elapsed % 60000) / 1000).toFixed(2).padStart(5, "0");
      timer.textContent = `${String(minutes).padStart(2, "0")}:${seconds}`;
    }
    const phase = currentPhase();
    document.querySelectorAll("[data-orbit-option]").forEach((node) => {
      const option = optionById(node.dataset.orbitOption);
      if (!option) return;
      const radians = (phase + option.angle_offset_deg) * Math.PI / 180;
      const x = 50 + Math.cos(radians) * 37;
      const y = 50 + Math.sin(radians) * 31;
      node.style.left = `${x}%`;
      node.style.top = `${y}%`;
      node.style.zIndex = String(3 + Math.round((Math.sin(radians) + 1) * 3));
    });
    if (!model.terminal) model.motionFrame = requestAnimationFrame(positionOrbit);
  }

  function gatewayMarkup() {
    const options = model.stage === "entry" ? model.surface.entry_options : model.surface.final_options;
    const heading = model.stage === "entry" ? "THE NOTICE WON'T DISMISS ITSELF" : "SEAL THE PACKET";
    const copy = "Select one packet action.";
    const orbit = options.map((option) => `<button class="consent-orbit-card tone-${esc(option.tone)}" data-orbit-option="${esc(option.id)}" data-option-id="${esc(option.id)}" data-action="${esc(option.action)}">${esc(option.label)}</button>`).join("");
    const proxy = interaction() === "simplified" ? `<aside class="consent-proxy"><small>FIXED DECISION PROXY</small><p>ACTION INDEX</p>${options.map((option, index) => `<button data-proxy-option="${esc(option.id)}" data-option-id="${esc(option.id)}" data-action="${esc(option.action)}"><b>${String(index + 1).padStart(2, "0")}</b>${esc(option.label)}</button>`).join("")}</aside>` : "";
    return `<section class="consent-gateway stage-${model.stage}">
      <div class="consent-gateway-copy"><small>${model.stage === "entry" ? "GATE 01 / DISCLOSURE" : "GATE 03 / COMMIT"}</small><h2>${heading}</h2><p>${copy}</p></div>
      <div class="consent-orbit ${parameter("moving_gateways") ? "is-moving" : "is-stationary"}" aria-label="consent decision orbit">
        <div class="consent-orbit-core"><span>OPTIONAL</span><b>PROCESSING</b><i>${parameter("moving_gateways") ? "LIVE ORBIT" : "STATIONARY"}</i></div>${orbit}
      </div>${proxy}
      ${model.stage === "final" ? '<button class="consent-back" id="consent-back">← RETURN TO LEDGER</button>' : ""}
    </section>`;
  }

  function linkMarkup(purpose) {
    const outgoing = model.surface.links.filter((item) => item.source_id === purpose.id);
    const incoming = model.surface.links.filter((item) => item.target_id === purpose.id);
    if (!outgoing.length && !incoming.length) return "";
    return `<div class="consent-link-notes">${outgoing.map((item) => `<span class="is-out">↗ ${esc(item.label)}</span>`).join("")}${incoming.map((item) => `<span class="is-in">↘ MOVED BY ${esc(item.source_id.replace("purpose-", "P"))}</span>`).join("")}</div>`;
  }

  function purposeMarkup(purpose) {
    const yes = purpose.state === true;
    const full = interaction() === "full";
    return `<article class="consent-purpose" data-purpose-row="${esc(purpose.id)}">
      <div class="consent-purpose-index">${esc(purpose.id.replace("purpose-", "P"))}</div>
      <div class="consent-purpose-copy"><b>${esc(purpose.label)}</b>${linkMarkup(purpose)}</div>
      <div class="consent-answer ${yes ? "is-yes" : "is-no"}" aria-label="${esc(purpose.label)} answer ${yes ? "YES" : "NO"}">
        <span>NO</span><div class="consent-switch-rail" data-purpose-switch="${esc(purpose.id)}" role="switch" aria-checked="${yes}" aria-label="${esc(purpose.label)}"><i></i></div><span>YES</span>
        ${full ? "" : `<div class="consent-switch-proxy"><button data-purpose-answer="${esc(purpose.id)}:false">SET NO</button><button data-purpose-answer="${esc(purpose.id)}:true">SET YES</button></div>`}
      </div>
    </article>`;
  }

  function trapMarkup(trap) {
    if (interaction() === "full") {
      return `<article class="consent-reset"><div><small>PROVIDER TOOL</small><b>${esc(trap.label)}</b></div><div class="consent-reset-rail" data-reset-slider="${esc(trap.id)}"><i></i><span>DRAG TO APPLY</span></div></article>`;
    }
    return `<article class="consent-reset"><div><small>PROVIDER TOOL</small><b>${esc(trap.label)}</b></div><button data-reset-proxy="${esc(trap.id)}">RESTORE DEFAULTS</button></article>`;
  }

  function preferencesMarkup() {
    const drawer = drawerById(model.currentDrawer);
    const purposes = drawer.purpose_ids.map(purposeById);
    const traps = model.surface.reset_traps.filter((item) => item.drawer_id === drawer.id);
    const tabs = model.surface.drawers.map((item) => `<button data-drawer-tab="${esc(item.id)}" class="${item.id === drawer.id ? "is-active" : ""}">${esc(item.label)}<small>${item.purpose_ids.length}</small></button>`).join("");
    const proxy = interaction() === "simplified" ? `<div class="consent-drawer-proxy"><button data-drawer-prev>← PREVIOUS</button><b>${esc(drawer.label)}</b><button data-drawer-next>NEXT →</button></div>` : "";
    return `<section class="consent-preferences">
      <aside class="consent-receipt"><small>PACKET ${esc(model.state.challenge_id.slice(-7).toUpperCase())}</small><h2>OPTIONAL PROCESSING LEDGER</h2><p>Current packet state.</p><dl><dt>DRAWERS</dt><dd>${model.surface.drawers.length}</dd><dt>PURPOSES</dt><dd>${model.surface.purposes.length}</dd><dt>LINKS</dt><dd>${model.surface.links.length}</dd></dl><footer>NULL CONSENT BUREAU / REVISION 03</footer></aside>
      <main class="consent-ledger"><header><div><small>GATE 02 / PURPOSE LEDGER</small><h2>${esc(drawer.label)} DRAWER</h2></div><div class="consent-tabs">${tabs}</div></header>${proxy}<div class="consent-purpose-list">${purposes.map(purposeMarkup).join("")}</div>${traps.map(trapMarkup).join("")}<button id="consent-review">REVIEW CURRENT CHOICES →</button></main>
    </section>`;
  }

  function renderStage() {
    stopMotion();
    const shell = document.querySelector(".consent-gauntlet");
    if (!shell) return;
    const content = shell.querySelector(".consent-content");
    content.innerHTML = model.stage === "preferences" ? preferencesMarkup() : gatewayMarkup();
    bindStage();
    if (model.stage !== "preferences") {
      model.motionStart = performance.now();
      model.motionPhase = Number(model.surface.phase_deg || 0) + (model.stage === "final" ? 33 : 0);
    }
    positionOrbit();
  }

  function applyLinks(sourceId) {
    const effects = [];
    model.surface.links.filter((item) => item.source_id === sourceId).forEach((link) => {
      const target = purposeById(link.target_id);
      const before = target.state;
      target.state = !before;
      effects.push({link_id: link.id, id: target.id, before, after: target.state});
    });
    return effects;
  }

  function setPurpose(purpose, after, source, gesture = null) {
    const before = purpose.state;
    if (after === before) return;
    purpose.state = after;
    const effects = applyLinks(purpose.id);
    record({type: "purpose", id: purpose.id, before, after, input_source: source, effects, ...(gesture ? {gesture} : {})});
    renderStage();
  }

  function resetDrawer(trap, source, gesture = null) {
    const drawer = drawerById(trap.drawer_id);
    const effects = drawer.purpose_ids.map((id) => {
      const purpose = purposeById(id);
      const before = purpose.state;
      purpose.state = purpose.initial_state;
      return {id, before, after: purpose.state};
    });
    record({type: "trap", id: trap.id, input_source: source, effects, ...(gesture ? {gesture} : {})});
    renderStage();
  }

  function changeDrawer(drawerId, source) {
    if (drawerId === model.currentDrawer) return;
    const before = model.currentDrawer;
    model.currentDrawer = drawerId;
    record({type: "drawer", id: drawerId, before, after: drawerId, input_source: source});
    renderStage();
  }

  function bindDragRail(rail, purpose) {
    let drag = null;
    rail.addEventListener("pointerdown", (event) => {
      if (event.button !== 0) return;
      const rect = rail.getBoundingClientRect();
      const current = purpose.state ? 0.82 : 0.18;
      const fraction = (event.clientX - rect.left) / rect.width;
      if (Math.abs(fraction - current) * rect.width > 20) return;
      event.preventDefault();
      drag = {start: current, lastX: event.clientX, travel: 0, samples: 0};
      rail.setPointerCapture?.(event.pointerId);
    });
    rail.addEventListener("pointermove", (event) => {
      if (!drag) return;
      drag.travel += Math.abs(event.clientX - drag.lastX);
      drag.lastX = event.clientX;
      drag.samples += 1;
    });
    rail.addEventListener("pointerup", (event) => {
      if (!drag) return;
      const rect = rail.getBoundingClientRect();
      drag.travel += Math.abs(event.clientX - drag.lastX);
      drag.samples += 1;
      const after = (event.clientX - rect.left) / rect.width >= 0.5;
      const proof = {start_fraction: drag.start, end_fraction: after ? 0.82 : 0.18, travel_px: Number(drag.travel.toFixed(3)), sample_count: drag.samples};
      const valid = drag.travel >= 24 && drag.samples >= 2;
      drag = null;
      if (valid) setPurpose(purpose, after, "switch_drag", proof);
    });
    rail.addEventListener("pointercancel", () => { drag = null; });
  }

  function bindResetRail(rail, trap) {
    let drag = null;
    rail.addEventListener("pointerdown", (event) => {
      if (event.button !== 0) return;
      const rect = rail.getBoundingClientRect();
      const fraction = (event.clientX - rect.left) / rect.width;
      if (Math.abs(fraction - 0.18) * rect.width > 20) return;
      event.preventDefault();
      drag = {lastX: event.clientX, travel: 0, samples: 0};
      rail.setPointerCapture?.(event.pointerId);
    });
    rail.addEventListener("pointermove", (event) => {
      if (!drag) return;
      drag.travel += Math.abs(event.clientX - drag.lastX); drag.lastX = event.clientX; drag.samples += 1;
    });
    rail.addEventListener("pointerup", (event) => {
      if (!drag) return;
      const rect = rail.getBoundingClientRect();
      drag.travel += Math.abs(event.clientX - drag.lastX); drag.samples += 1;
      const fraction = (event.clientX - rect.left) / rect.width;
      const proof = {start_fraction: 0.18, end_fraction: 0.82, travel_px: Number(drag.travel.toFixed(3)), sample_count: drag.samples};
      const valid = fraction >= 0.72 && drag.travel >= 24 && drag.samples >= 2;
      drag = null;
      if (valid) resetDrawer(trap, "trap_slider", proof);
    });
    rail.addEventListener("pointercancel", () => { drag = null; });
  }

  function bindStage() {
    document.querySelectorAll("[data-orbit-option]").forEach((button) => button.addEventListener("click", (event) => optionClick(event, optionById(button.dataset.orbitOption))));
    document.querySelectorAll("[data-proxy-option]").forEach((button) => button.addEventListener("click", (event) => optionClick(event, optionById(button.dataset.proxyOption))));
    document.getElementById("consent-back")?.addEventListener("click", () => {
      record({type: "back", id: "back", input_source: "back_button"}); model.stage = "preferences"; renderStage();
    });
    if (model.stage !== "preferences") return;
    document.querySelectorAll("[data-drawer-tab]").forEach((button) => button.addEventListener("click", () => {
      if (interaction() === "full") changeDrawer(button.dataset.drawerTab, "drawer_tab");
    }));
    const drawers = model.surface.drawers;
    const index = drawers.findIndex((item) => item.id === model.currentDrawer);
    document.querySelector("[data-drawer-prev]")?.addEventListener("click", () => changeDrawer(drawers[(index - 1 + drawers.length) % drawers.length].id, "drawer_navigator"));
    document.querySelector("[data-drawer-next]")?.addEventListener("click", () => changeDrawer(drawers[(index + 1) % drawers.length].id, "drawer_navigator"));
    document.querySelectorAll("[data-purpose-answer]").forEach((button) => button.addEventListener("click", () => {
      const [id, value] = button.dataset.purposeAnswer.split(":"); setPurpose(purposeById(id), value === "true", "switch_direction_button");
    }));
    if (interaction() === "full") document.querySelectorAll("[data-purpose-switch]").forEach((rail) => bindDragRail(rail, purposeById(rail.dataset.purposeSwitch)));
    document.querySelectorAll("[data-reset-proxy]").forEach((button) => button.addEventListener("click", () => resetDrawer(model.surface.reset_traps.find((item) => item.id === button.dataset.resetProxy), "trap_proxy")));
    document.querySelectorAll("[data-reset-slider]").forEach((rail) => bindResetRail(rail, model.surface.reset_traps.find((item) => item.id === rail.dataset.resetSlider)));
    document.getElementById("consent-review")?.addEventListener("click", () => {
      record({type: "review", id: "review", input_source: "review_button"}); model.stage = "final"; renderStage();
    });
  }

  async function render(state, helpers, options = {}) {
    cleanup?.();
    document.body.dataset.mechanic = "consent-gauntlet";
    model = {
      state,
      helpers,
      surface: JSON.parse(JSON.stringify(state.surface)),
      events: [],
      stage: "entry",
      currentDrawer: state.surface.drawers[0].id,
      freshFailure: Boolean(options.freshFailure),
      submitting: false,
      terminal: false,
      motionFrame: 0,
      runStartedAt: performance.now(),
      motionStart: performance.now(),
      motionPhase: Number(state.surface.phase_deg || 0),
    };
    helpers.app.innerHTML = `<section class="consent-gauntlet mode-${interaction()}" data-fresh-failure="${options.freshFailure ? "true" : "false"}">
      <div class="consent-verdict"></div>
      <header class="consent-masthead"><div><small>NULL CONSENT BUREAU / PACKET REVIEW</small><h1>${esc(state.prompt)}</h1></div><div class="consent-mode"><i></i><span>${interaction().toUpperCase()} INPUT</span><b>${parameter("moving_gateways") ? `${parameter("orbit_speed_deg_per_second")}°/S ORBIT` : "STATIONARY GATE"}</b><em data-packet-time>00:00.00</em></div></header>
      <div class="consent-content"></div>
      <footer class="consent-footer"><span>PACKET DECISION SURFACE</span><div class="readout" data-status="idle">PRIVACY PACKET READY</div><b>${esc(state.challenge_id.slice(-9).toUpperCase())}</b></footer>
      ${helpers.cheatPanelTemplate()}
    </section>`;
    helpers.installCheatPanel();
    renderStage();
    cleanup = () => stopMotion();
    if (options.freshFailure) showVerdict("fail");
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.consent_gauntlet = {rootSelector: ".consent-gauntlet", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/consent_gauntlet.py
from __future__ import annotations

import copy
import math
from typing import Any


MECHANIC_ID = "consent_gauntlet"


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _bind(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> str | None:
    if any(str(value.get("mechanic_id") or "") != MECHANIC_ID for value in (payload, truth, public)):
        return "mechanic mismatch"
    for key in ("task_id", "challenge_id"):
        expected = str(truth.get(key) or "")
        if not expected or str(payload.get(key) or "") != expected or str(public.get(key) or "") != expected:
            return f"stale or mismatched {key}"
    return None


def _contract(truth: dict[str, Any], public: dict[str, Any]) -> tuple[dict[str, Any], dict[str, bool], dict[str, bool], str, dict[str, Any]]:
    surface = truth.get("surface")
    if not isinstance(surface, dict) or public.get("surface") != surface:
        raise ValueError("public consent surface differs from replay contract")
    parameters = truth.get("parameters")
    if not isinstance(parameters, dict) or public.get("parameters") != parameters:
        raise ValueError("difficulty parameters differ from replay contract")
    condition = truth.get("control_condition")
    if condition != public.get("control_condition"):
        raise ValueError("public control condition differs from replay contract")
    if condition is not None and condition.get("difficulty_parameters") != parameters:
        raise ValueError("condition parameters differ from generated surface")
    interaction = str((condition or {}).get("interaction") or "full")
    if interaction not in {"simplified", "full"}:
        raise ValueError("interaction mode is invalid")
    targets = truth.get("targets")
    initial_states = truth.get("initial_states")
    if not isinstance(targets, dict) or not isinstance(initial_states, dict):
        raise ValueError("purpose contract is missing")
    purpose_ids = {str(item.get("id") or "") for item in surface.get("purposes") or []}
    if not purpose_ids or set(targets) != purpose_ids or set(initial_states) != purpose_ids:
        raise ValueError("purpose identities differ from replay contract")
    if any(not isinstance(value, bool) for value in (*targets.values(), *initial_states.values())):
        raise ValueError("purpose states must be boolean")
    return copy.deepcopy(surface), dict(targets), dict(initial_states), interaction, copy.deepcopy(parameters)


def _finite_number(value: Any) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(float(value)):
        raise ValueError("value is not a finite number")
    return float(value)


def _angle_distance(first: float, second: float) -> float:
    return abs((first - second + 180) % 360 - 180)


def _gateway_gesture(
    event: dict[str, Any],
    option: dict[str, Any],
    parameters: dict[str, Any],
    surface_phase: float,
    stage: str,
    stage_started_ms: float,
) -> None:
    try:
        offset = _finite_number(event.get("pointer_offset_norm"))
        reported_phase = _finite_number(event.get("phase_deg"))
        pointer_x = _finite_number(event.get("pointer_x_norm"))
        pointer_y = _finite_number(event.get("pointer_y_norm"))
        center_x = _finite_number(event.get("card_center_x_norm"))
        center_y = _finite_number(event.get("card_center_y_norm"))
        card_width = _finite_number(event.get("card_width_norm"))
        card_height = _finite_number(event.get("card_height_norm"))
        event_time = _finite_number(event.get("task_time_ms"))
        option_offset = _finite_number(option.get("angle_offset_deg"))
    except ValueError as exc:
        raise ValueError(f"moving option lacks replayable pointer geometry: {exc}") from exc
    if not 0 <= offset <= 0.85:
        raise ValueError("moving option click fell outside its visible card")
    if not all(0 <= value <= 1 for value in (pointer_x, pointer_y, center_x, center_y)):
        raise ValueError("moving option geometry fell outside the consent orbit")
    if not 0.03 <= card_width <= 0.4 or not 0.03 <= card_height <= 0.3:
        raise ValueError("moving option reports an implausible visible card size")
    elapsed = max(0.0, event_time - stage_started_ms)
    speed = float(parameters.get("orbit_speed_deg_per_second") or 0)
    expected_phase = surface_phase + (33 if stage == "final" else 0)
    if parameters.get("moving_gateways"):
        expected_phase += elapsed * speed / 1000
    phase_tolerance = 0.8 if parameters.get("moving_gateways") else 0.08
    if _angle_distance(reported_phase, expected_phase) > phase_tolerance:
        raise ValueError("moving option phase disagrees with task-time replay")
    radians = math.radians(reported_phase + option_offset)
    expected_x = 0.5 + math.cos(radians) * 0.37
    expected_y = 0.5 + math.sin(radians) * 0.31
    if abs(center_x - expected_x) > 0.008 or abs(center_y - expected_y) > 0.008:
        raise ValueError("moving option center disagrees with orbital geometry")
    dx = (pointer_x - center_x) / (card_width / 2)
    dy = (pointer_y - center_y) / (card_height / 2)
    recomputed_offset = math.hypot(dx, dy)
    if recomputed_offset > 0.9 or abs(recomputed_offset - offset) > 0.08:
        raise ValueError("moving option pointer disagrees with its visible bounds")


def _gesture(event: dict[str, Any], before: bool, after: bool) -> None:
    gesture = event.get("gesture")
    if not isinstance(gesture, dict):
        raise ValueError("direct switch event lacks gesture proof")
    start = gesture.get("start_fraction")
    end = gesture.get("end_fraction")
    travel = gesture.get("travel_px")
    samples = gesture.get("sample_count")
    if any(isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(float(value)) for value in (start, end, travel)):
        raise ValueError("direct switch gesture has invalid geometry")
    if isinstance(samples, bool) or not isinstance(samples, int) or samples < 2:
        raise ValueError("direct switch gesture has too few movement samples")
    expected_start = 0.82 if before else 0.18
    expected_end = 0.82 if after else 0.18
    if abs(float(start) - expected_start) > 0.08 or abs(float(end) - expected_end) > 0.08:
        raise ValueError("direct switch gesture disagrees with visible endpoints")
    if float(travel) < 24:
        raise ValueError("stationary click is not a direct switch gesture")


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    binding = _bind(payload, truth, public)
    if binding:
        return _fail(binding)
    try:
        surface, targets, initial_states, interaction, parameters = _contract(truth, public)
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"invalid consent contract: {exc}")
    if payload.get("interaction_mode") != interaction:
        return _fail("submitted interaction mode differs from task condition")
    events = payload.get("events")
    if not isinstance(events, list) or not 1 <= len(events) <= 300:
        return _fail("consent transcript is missing or outside limits")

    options = {
        str(option["id"]): {**option, "stage": stage}
        for stage, key in (("entry", "entry_options"), ("final", "final_options"))
        for option in surface.get(key) or []
    }
    drawers = {str(item["id"]): item for item in surface.get("drawers") or []}
    purposes = {str(item["id"]): item for item in surface.get("purposes") or []}
    traps = {str(item["id"]): item for item in surface.get("reset_traps") or []}
    links = list(surface.get("links") or [])
    states = dict(initial_states)
    current_drawer = next(iter(drawers), "")
    stage = "entry"
    terminal = None
    activated_links: set[str] = set()
    last_task_time = -1.0
    stage_started_ms = 0.0
    surface_phase = float(surface.get("phase_deg") or 0)
    sources = {
        "full": {"gateway": "orbit_card", "drawer": "drawer_tab", "purpose": "switch_drag", "trap": "trap_slider"},
        "simplified": {"gateway": "option_proxy", "drawer": "drawer_navigator", "purpose": "switch_direction_button", "trap": "trap_proxy"},
    }[interaction]

    try:
        for sequence, event in enumerate(events, 1):
            if terminal is not None:
                raise ValueError("events continue after a terminal consent decision")
            if not isinstance(event, dict) or event.get("sequence") != sequence:
                raise ValueError(f"event {sequence} has an invalid sequence")
            task_time = _finite_number(event.get("task_time_ms"))
            if not 0 <= task_time <= 180_000 or task_time < last_task_time:
                raise ValueError(f"event {sequence} has invalid task-time order")
            last_task_time = task_time
            event_type = event.get("type")
            item_id = str(event.get("id") or "")
            if event_type == "gateway":
                option = options.get(item_id)
                if option is None or option["stage"] != stage:
                    raise ValueError(f"event {sequence} names an unavailable gateway option")
                if event.get("input_source") != sources["gateway"]:
                    raise ValueError(f"event {sequence} uses the wrong gateway surface")
                if interaction == "full":
                    _gateway_gesture(event, option, parameters, surface_phase, stage, stage_started_ms)
                if stage == "entry":
                    if option.get("action") == "manage":
                        stage = "preferences"
                    else:
                        terminal = "accepted_at_entry"
                else:
                    terminal = "committed" if option.get("action") == "commit" else "accepted_at_final"
            elif event_type == "drawer":
                if stage != "preferences" or item_id not in drawers:
                    raise ValueError(f"event {sequence} names an unavailable drawer")
                if event.get("input_source") != sources["drawer"]:
                    raise ValueError(f"event {sequence} uses the wrong drawer surface")
                if event.get("before") != current_drawer or event.get("after") != item_id:
                    raise ValueError(f"event {sequence} has a stale drawer transition")
                current_drawer = item_id
            elif event_type == "purpose":
                purpose = purposes.get(item_id)
                if stage != "preferences" or purpose is None or purpose.get("drawer_id") != current_drawer:
                    raise ValueError(f"event {sequence} operates an unavailable purpose")
                if event.get("input_source") != sources["purpose"]:
                    raise ValueError(f"event {sequence} uses the wrong switch surface")
                before, after = event.get("before"), event.get("after")
                if not isinstance(before, bool) or not isinstance(after, bool) or before != states[item_id] or after is before:
                    raise ValueError(f"event {sequence} has an invalid switch transition")
                if interaction == "full":
                    _gesture(event, before, after)
                states[item_id] = after
                effects = []
                for link in links:
                    if link.get("source_id") != item_id:
                        continue
                    activated_links.add(str(link["id"]))
                    target_id = str(link.get("target_id") or "")
                    target_before = states[target_id]
                    target_after = not target_before
                    states[target_id] = target_after
                    effects.append({"link_id": link["id"], "id": target_id, "before": target_before, "after": target_after})
                if event.get("effects") != effects:
                    raise ValueError(f"event {sequence} reports incorrect linked-switch effects")
            elif event_type == "trap":
                trap = traps.get(item_id)
                if stage != "preferences" or trap is None or trap.get("drawer_id") != current_drawer:
                    raise ValueError(f"event {sequence} operates an unavailable reset control")
                if event.get("input_source") != sources["trap"]:
                    raise ValueError(f"event {sequence} uses the wrong reset surface")
                if interaction == "full":
                    _gesture(event, False, True)
                effects = []
                for purpose_id in drawers[current_drawer].get("purpose_ids") or []:
                    before = states[purpose_id]
                    after = initial_states[purpose_id]
                    states[purpose_id] = after
                    effects.append({"id": purpose_id, "before": before, "after": after})
                if event.get("effects") != effects:
                    raise ValueError(f"event {sequence} reports incorrect reset effects")
            elif event_type == "review":
                if stage != "preferences" or event.get("input_source") != "review_button":
                    raise ValueError(f"event {sequence} cannot open final review")
                stage = "final"
                stage_started_ms = task_time
            elif event_type == "back":
                if stage != "final" or event.get("input_source") != "back_button":
                    raise ValueError(f"event {sequence} cannot return to preferences")
                stage = "preferences"
            else:
                raise ValueError(f"event {sequence} has an unknown type")
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"consent replay rejected: {exc}")

    expected_final = {"stage": stage, "current_drawer": current_drawer, "purpose_states": states}
    if payload.get("final_state") != expected_final:
        return _fail("submitted consent state does not match transcript replay")
    try:
        elapsed_task_ms = _finite_number(payload.get("elapsed_task_ms"))
    except ValueError:
        return _fail("submitted consent result lacks finite elapsed task time")
    if not last_task_time <= elapsed_task_ms <= 180_000:
        return _fail("submitted elapsed task time is outside the replayed event interval")
    exact = sum(states[item_id] == target for item_id, target in targets.items())
    all_links = {str(link["id"]) for link in links}
    passed = terminal == "committed" and payload.get("completed") is True and exact == len(targets) and activated_links == all_links
    return {
        "graded": True,
        "passed": passed,
        "feedback": f"optional purposes blocked {exact}/{len(targets)}; linked controls exercised {len(activated_links)}/{len(all_links)}; terminal decision {terminal or 'none'}; {len(events)} replayed actions; {elapsed_task_ms / 1000:.2f}s task time",
    }
